packages/devtools_app/lib/src/screens/memory/shared/heap/heap.dart

```diff
@@ -29,17 +35,62 @@ class AdaptedHeap {
       singleHeapClass.countInstance(data, i);
     }
 
-    return SingleHeapClasses(result)..seal();
+    return SingleHeapClasses(result, filter)..seal();
   }
 }
 
-abstract class HeapClasses with Sealable {
-  Iterable<ClassStats> get classStatsList;
+abstract class HeapClasses<T extends ClassStats> with Sealable {
+  HeapClasses(this.filter);
+
+  List<T> get classStatsList;
+
+  final ValueListenable<ClassFilter> filter;
+}
+
+mixin Filterable<T extends ClassStats> on HeapClasses<T> {
+  ClassFilter? _appliedFilter;
+  List<T>? _filtered;
+
+  List<T> filtered() {
+    final newFilter = filter.value;
+    final oldFilter = _appliedFilter;
+    final oldFiltered = _filtered;
+    _appliedFilter = newFilter;
+    if ((oldFilter == null) != (oldFiltered == null)) {
+      throw StateError('nullness should match');
```

Comment:
this error message might need a bit more context.

Reply:
I assume the person who investigate it will read the code. And code is very descriptive here.
Do see cases when code will not be available for the investigator?